### ISSUE TITLE
CPK breaks preloading of non-CPK polymorphic associations

### ISSUE THREAD

**Issue by MartinKoerner** [composite-primary-keys/composite_primary_keys#369]
Hi,

with Rails 4.2.7.1 we have included CPK in the bundle file, but (for testing reasons) NO model or association with composite primary keys.

We have model with following association

    class SomeModel < ActiveRecord::Base
      has_and_belongs_to_many :specific_uploads, -> { where('type = ?', 'SpecificUpload') },
                              class_name: 'Upload',
                              join_table: '...'
    end

If I write a query like
  `SomeModel.where(....).preload(:specific_uploads)`
I get a very strange result: The specific upload for the first found record is loaded, but all others are replaced by an empty array. No error is thrown. It just looks like, there are no uploads on these models (while there are some in the DB).

If I remove the polymorphic condition `-> { where('type = ?', 'SpecificUpload') }`, it works correctly again.

**Comment by MartinKoerner** [comment]
I began debugging this and the problem got even weirder. I started my search in `associated_records_by_owner`in `preloader/associations.rb`.
The owners are loaded correctly, the sliced IDs also are correct. In my case, I only have one slice with 3 IDs.

Now, the `records_for` method is called with these 3 IDs as one array. It should return 6 Elements (1 Element for ID 1, 3 Elements for ID 2, 2 Elements for ID 3). But it only returns the first one.
I checked the development logs, and the used Query correctly returns all 6 elements, if fired directly. It is the exact same query, if I debug the code without CPK.

`records_for`is implemented in the original AR code with `scope.where(association_key.in(ids))`- and the Scope itself looks OK (it also is supported by the fact, that the generated query is OK). It is not overridden by CPK.

So I guess, the problem lies somewhere between the execution of the (correct) query, and the return value of `.where`on the association scope. ATM I'm out of ideas

**Comment by cfis** [comment]
Hmm, do you have a test case?  Particularly one with the test data that ships with CPK?

**Comment by MartinKoerner** [comment]
The test data from CPK can't be loaded into mysql/sqlite3, because of some definition errors (it's broken on travis as well). I'll try to put an example with a test case up tomorrow.

**Comment by cfis** [comment]
Sqlite and mysql should be fixed on master now.

**Comment by MartinKoerner** [comment]
Hi, I added some test cases in my branch (created from the 4.2.x-branch). I got slightly different results than in my description, but nevertheless, they are wrong :)

It seems, that has_many associations don't have a problem, while habtm associations do, even without preloading. My tests show both failures for
- one record got loaded twice instead of once
- one record is missing in the result set

Hope, these cases help

**Comment by cfis** [comment]
Thanks for the tests.  I'll add them to the master branch and see if I can reproduce the problem and fix.  Do you have plans to move to Rails 5?

**Comment by cfis** [comment]
The issue is the association is setup incorrectly.  What you want is this:

has_and_belongs_to_many :groups_2, class_name: 'Group', foreign_key: 'employee_id', join_table: 'employees_groups'

Notice the foreign_key is employee_id, not group_id.

Anyway, I did add these tests to master branch.

Thanks for the help on this.

**Comment by MartinKoerner** [comment]
@cfis Ok, that's on me :(
As my original bug contained models using STI (which my example test cases did not), I created a simple Rails project with one model and a HABTM association to another one with two derivations Type1 and Type2.

[cpk_bug_recreate.zip](https://github.com/composite-primary-keys/composite_primary_keys/files/680097/cpk_bug_recreate.zip)
There, the preloading test fails with Rails 4.2.7.1 and CPK 8.0.1 - I attached the project for reference (it should be ready to run with just `bundle install; ruby test/relation_test.rb`). Maybe you want to adapt it to your test data?

The same test works correctly with Rails 5.0.1 and CPK 9.0.4, so in my opinion, the ticket can stay closed.
